tiledb/sm/array/array.cc

```diff
@@ -876,25 +941,42 @@ std::vector<shared_ptr<const Enumeration>> Array::get_enumerations(
 
     // Store the loaded enumerations in the schema
     for (auto& enmr : loaded) {
-      schema->store_enumeration(enmr);
+      opened_array_->array_schema_latest_ptr()->store_enumeration(enmr);
     }
   }
 
   // Return the requested list of enumerations
   std::vector<shared_ptr<const Enumeration>> ret(enumeration_names.size());
   for (size_t i = 0; i < enumeration_names.size(); i++) {
-    ret[i] = schema->get_enumeration(enumeration_names[i]);
+    ret[i] = array_schema_latest().get_enumeration(enumeration_names[i]);
   }
   return ret;
 }
 
-void Array::load_all_enumerations() {
+void Array::load_all_enumerations(bool all_schemas) {
   if (!is_open_) {
     throw ArrayException("Unable to load all enumerations; Array is not open.");
   }
   // Load all enumerations, discarding the returned list of loaded enumerations.
-  for (const auto& schema : array_schemas_all()) {
-    get_enumerations(schema.second->get_enumeration_names(), schema.second);
+  if (all_schemas) {
+    // Since array open v1 does not initialize array_schemas_all, we need to
+    // either serialize the full schemas with LoadEnumerationsResponse, or
+    // reopen the array using array open v2 (only if the config is set to use
+    // v1) so that we can store the enumerations from the response in the
+    // correct schemas using the schema names.
+
+    // For now, reopen the array if it's found to be using array open v1.
+    // Once we update to use post_array_schema_from_rest we will always have
+    // array_schemas_all initialized and this could go away.
```

Comment:
Just want to point these comments out. As I mentioned in the comment I could serialize allArraySchemas with the enumerations intact and that would be fine, but that would only be required [until post_array_schema_from_rest is finished](https://github.com/TileDB-Inc/TileDB/pull/5181/files#diff-74ce8c89d5884c4961474244d6263b78237ca72ec9e61c9a8268b4fab73b0b53R333-R339). Array open v3 is now the default so I thought this was better than locking us into serializing full schemas here.

If we want to just serialize all the schemas instead I can do that.

Reply:
Let's not do array operations under the hood. I have a more radical idea. How about returning a message at the C API level that this operation is unsupported for array v1 opened arrays and redirect the user to use v2 instead?

@ihnorton Are we ok to introduce new C APIs assuming array v2/query v3 now that it's default, and just be backwards compatible forever for the pre-existing ones?

Reply:
Done, I added exceptions in both the C API and in `tiledb::sm::Array::load_all_enumerations` in case it is called internally like we do in `unit-enumerations.cc`. I modified the existing tests to check for both exceptions if we are not using array open v2. The exception could be removed and tests updated as part of #5181.